### ISSUE TITLE
Suphx-style reward shaping

### ISSUE THREAD

**Issue by sotetsuk** [mjx-project/mjx#1127]

**Comment by sotetsuk** [comment]
`workspace/suphx-reward-shaping` の下で作業 @nissymori

**Comment by nissymori** [comment]
#### 特徴量  suphnx論文より
- 持ち点
- 上がり点/失点
-  親
- 本場
- 詰み棒

suphnxでは上記のの情報をGRUに渡しているつまりj局目の場合j組ある.
slackでの議論のように現在の情報のみから予測したい.

GRUでは過去の情報の系列から局数, 自風なども推論可能だが一局のみから予測する場合は明示的に特徴量を用意した方がよさそう.

#### 案
- 局終了時の持ち点
- 親
- 本場
- 詰み棒
- 自風
- 局

**Comment by nissymori** [comment]
### jaxで学習したモデルをnumpyへ吐き出せるか確認

簡単な例を作って確認.
https://colab.research.google.com/drive/1Hn5TADrD7NUHjxo0QqfSUQ8UX_XuPKId?usp=sharing

学習が終わったモデルの重みを 
`jnp.save("weights.npy", weights)`で保存
`jnp.load("weights.npy")`でロードできる.
loadしたobjectのtypeを確認すると`numpy.ndarray`なので少なくともこの方法でjaxで学習させた重みをnumpyのarrayとして使える.

**Comment by nissymori** [comment]
特徴量 | 前処理方法

- 自身の点 | ÷ 100000
- [上家, 下家, 対面] 自分の点との差 |  ÷ 100000
- 自風: one-hot
- 親: one-hot
- 本場: 整数 |  ÷ 4 適当
- 詰み棒: 整数 |  ÷ 4 適当
- 局: 整数 |   ÷ 8 適当

**Comment by sotetsuk** [comment]
上家とかの並びにするのは重要な気がするけどなあ

**Comment by nissymori** [comment]
> 上家とかの並びにするのは重要な気がするけどなあ

単純に4人の点/100000だと全くうまくいかないのでこちらに変えます！

**Comment by nissymori** [comment]
変更点
- 特徴量として残りの親の数を追加する.
- 風の情報を
- 出力は4次元[起家, ..ラス親]
- NNは8個用意する[model_0, ...model_7](東1から南4)

procedure
-  南4のデータから順位点をmodel_7に予測させる
- 南３のデータのmodel_7の予測結果をmodel_6に予測させる.
- 以下繰り返す.


ToDo

- [x] 特徴量実装
- [x] 特徴量テスト
- [ ] 訓練関数実装
- [ ] 訓練関数テスト
- [ ] plot関数実装
- [ ] plot関数テスト
- [ ] コマンドライン整備
- [ ] 実験
